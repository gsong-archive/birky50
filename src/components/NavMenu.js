import React from "react";

import { Nav, NavItem, NavLink } from "reactstrap";

export default ({ sections, onClick }) => {
  const navItems = sections.map(({ id, navLabel, LabelComponent }) => (
    <NavItem key={id}>
      <NavLink href={`#${id}`} onClick={onClick(id)}>
        <LabelComponent label={navLabel} />
      </NavLink>
    </NavItem>
  ));

  return <Nav fill={true}>{navItems}</Nav>;
};
