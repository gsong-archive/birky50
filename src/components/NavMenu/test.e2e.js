import { screenshot, settings } from "testUtils/puppeteer";

beforeAll(async () => {
  jest.setTimeout(settings.jestTimeOut);
  await page.goto(settings.appUrl);
});

describe("Click item in nav menu should scroll to section", () => {
  const sections = {
    Details: "Celebration Details",
    Airport: "Where to Fly Into",
    Lodging: "Where to Stay",
    Food: "Where to Eat",
    Activities: "What to Do",
    Questions: "Questions",
  };

  Object.entries(sections).forEach(([navLabel, sectionLabel], i) => {
    test(navLabel, async () => {
      const link = await getByText(page, "nav//*", navLabel);
      const section = await getByText(page, "h1//*", sectionLabel);

      // 1. Click on nav menu link
      // 2. Wait for the screen to stop scrolling
      // 3. Make sure the section associated with the link is visible
      await link.click();
      await stoppedScrolling(section);
      await isVisibleInViewPort(section);

      await screenshot(`${i}-${navLabel}`, page);
    });
  });
});

const stoppedScrolling = async element => {
  let boundingBox = await element.boundingBox();
  let stillScrolling = true;

  while (stillScrolling) {
    await delay(settings.scrollCheckDelay);
    const oldBoundingBox = boundingBox;
    boundingBox = await element.boundingBox();
    if (oldBoundingBox.y === boundingBox.y) stillScrolling = false;
  }
};

const isVisibleInViewPort = async element => {
  const boundingBox = await element.boundingBox();

  expect(boundingBox.y).toBeGreaterThan(0);
  expect(boundingBox.y + boundingBox.height).toBeLessThan(
    page.viewport().height
  );
};

const getByText = async (container, selector, text) => {
  const elements = await container.$x(
    `//${selector}[text()[contains(., '${text}')]]`
  );

  return elements[0];
};

const delay = timeout => new Promise(resolve => setTimeout(resolve, timeout));
