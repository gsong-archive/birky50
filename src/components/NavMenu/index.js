//@flow strict
import React from "react";
import styled from "react-emotion";

import { Link } from "../../styles/components";

import type { SectionData } from "../Section.type";

type Props = {
  sections: SectionData,
  onClick: string => (SyntheticMouseEvent<HTMLLinkElement>) => void,
};

export default ({ sections, onClick }: Props) => {
  const navItems = sections.map(({ id, navLabel, LabelComponent }, i) => (
    <NavItem key={id}>
      <NavLink href={`#${id}`} onClick={onClick(id)}>
        <LabelComponent label={navLabel} />
      </NavLink>
    </NavItem>
  ));

  return (
    <nav>
      <NavContainer>{navItems}</NavContainer>
    </nav>
  );
};

const NavContainer = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  list-style-type: none;
  padding: 0;
  margin: 0;
`;

const NavItem = styled.li`
  flex-grow: 1;
  text-align: center;
`;

const NavLink = styled(Link)`
  display: block;
  font-size: 1.15rem;
  padding: 0.75rem 1rem;
`;
