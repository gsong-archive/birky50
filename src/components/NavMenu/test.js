import React from "react";

import { render } from "react-testing-library";

import NavMenu from "./index";

test("NavMenu renders properly", () => {
  const props = {
    sections: [
      {
        id: "id",
        navLabel: "Menu Item",
        LabelComponent: props => props.label,
      },
    ],
    onClick: jest.fn(),
  };
  const { container } = render(<NavMenu {...props} />);

  expect(container.firstChild).toMatchSnapshot();
});
