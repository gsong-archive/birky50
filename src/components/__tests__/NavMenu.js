import React from "react";

import { render } from "react-testing-library";

import NavMenu from "../NavMenu";

test("NavMenu renders properly", () => {
  const props = {
    sections: [
      {
        id: "id",
        navLabel: "label",
        LabelComponent: props => <div>{props.label}</div>
      }
    ],
    onClick: jest.fn()
  };
  const { container } = render(<NavMenu {...props} />);

  expect(container.firstChild).toMatchSnapshot();
});
