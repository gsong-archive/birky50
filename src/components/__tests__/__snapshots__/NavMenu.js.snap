// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu renders properly 1`] = `
.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.emotion-2 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: center;
}

.emotion-0 {
  color: maroon;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
  padding: 0.75rem 1rem;
}

<nav>
  <ul
    class="emotion-4 emotion-5"
  >
    <li
      class="emotion-2 emotion-3"
    >
      <a
        href="#id"
        class="emotion-0 emotion-1"
      >
        Menu Item
      </a>
    </li>
  </ul>
</nav>
`;
